gms/endpoint_state.cc

```diff
@@ -61,6 +62,32 @@ future<> i_endpoint_state_change_subscriber::on_application_state_change(inet_ad
     return make_ready_future<>();
 }
 
+locator::host_id endpoint_state::get_host_id() const {
+    if (auto* app_state = get_application_state_ptr(application_state::HOST_ID)) {
+        return locator::host_id(utils::UUID(app_state->value()));
+    }
+    return locator::host_id::create_null_id();
```

Comment:
Is null id the best choice here? It's too easy for it to slip through and get used as if it is a real ID. Shouldn't we preserve it as optional so it's clear that it's unser?

Reply:
I don't think that's required, a null host_id is never valid and it generally stands for an absent host_id.
returning `std::optional<locator::host_id>` is an over-kill IMO.
The system would behave practically the same is std::nullopt is returned, or a std::optional<host_id> holding a null id is returned.  Both cases are invalid and shouldn't happen.

Reply:
A slight inconsistency with `get_tokens()`, which could return an empty set instead of nullopt following the same reasoning.

Reply:
In a previous version it did just that, but Avi asked to return nullopt to be able to distinguish the two cases.
In theory, we may want to keep a node with TOKENS application state containing an empty list (maybe when we deprecate vndoes entirely and rely only on tablets), but a node cannot have a null host_id.

Reply:
correction: we don't allow empty tokens list (see `system_keyspace::load_topology_state`)

Reply:
@avikivity, based on https://github.com/scylladb/scylladb/blob/57c408ab5d206980fece198a1278e5612629bf08/db/system_keyspace.cc#L2706-L2710, we cannot have empty tokens in system.peers, so we don't really need to return an optional set of tokens in endpoint_state::get_tokens, we can safely return an empty set instead if the application state is missing.